File: src/micapolos/SocksGirl.java
package micapolos;

import micapolos.zexy.Direction;
import micapolos.zexy.Key;

import static micapolos.zexy.Anchor.*;
import static micapolos.zexy.Image.*;
import static micapolos.zexy.Position.*;
import static micapolos.zexy.Sprite.*;
import static micapolos.zexy.Stack.*;

public class SocksGirl {
  static int imageIndex(Direction direction) {
    return switch (direction) {
      case UP -> 8;
      case UP_RIGHT -> 7;
      case RIGHT -> 6;
      case DOWN_RIGHT -> 5;
      case DOWN -> 0;
      case DOWN_LEFT -> 11;
      case LEFT -> 10;
      case UP_LEFT -> 9;
    };
  }

  static void main() {
    var sheetImages = image(SocksGirl.class, "socksgirl-sheet.png").sliceVertically(12);
    var groundImage = newImage("ground", 500, 256);
    var canvas = groundImage.newCanvas();
    canvas.fillRect(0, 200, 500, 56);
    canvas.fillTriangle(200, 200, 500, 200, 500, 100);
    var groundSprite = newSprite()
      .with(groundImage)
      .with(anchor(256, 200))
      .with(position(256, 128));

    var image = Direction
      .fromSpans(Key.LEFT.pressedSpan(), Key.RIGHT.pressedSpan(), Key.UP.pressedSpan(), Key.DOWN.pressedSpan())
      .orIfNull(Direction.DOWN)
      .mapToInteger(SocksGirl::imageIndex)
      .selectFrom(sheetImages);

    var girlSprite = newSprite()
      .withImage(image)
      .with(anchor(32, 64))
      .with(position(160, 128));

    stackOf(groundSprite, girlSprite).show();
  }
}

File: src/micapolos/zexy/examples/SocksGirl.kt
package micapolos.zexy.examples

import micapolos.zexy.*

fun main() {
  show {
    val direction = variable(4)
    val isTalking = key.z.isPressed
    val isMoving = key.x.isPressed
    val frameIndex = variable(0)

    val directionBaseIndex = direction.selectFrom(0, 7, 12, 17, 22, 27, 32, 37)
    val directionFrameCount = direction.selectFrom(6, 4, 4, 4, 4, 4, 4, 4)

    val spriteIndex = isTalking
      .ifTrue(frameIndex)
      .orElse(
        4.value + directionBaseIndex +
            isMoving
              .ifTrue(frameIndex + 1)
              .orElse(0)
      )

    val spriteFrameCount =
      isTalking
        .ifTrue(4)
        .orElse(
          isMoving
            .ifTrue(directionFrameCount)
            .orElse(1)
        )

    val girlSize = size(64, 64)

    repeat {
      frameIndex set (frameIndex + 1) % spriteFrameCount
      this pause 0.15
    }

    startOn(key.z.press) {
      direction set 4
    }

    startOn(key.right.press) {
      direction set (direction + 1) % 8
    }

    startOn(key.left.press) {
      direction set (direction + 7) % 8
    }

    sprite
      .with(image("/micapolos/socksgirl-sheet.png"))
      .with(position((screen.size.width - 64) / 2, 160))
      .with(girlSize)
      .withImage(position(spriteIndex * girlSize.width, 0))
  }
}